ple/games/tetris_engine.py

```diff
@@ -0,0 +1,518 @@
+            if not self.isValidPosition():
+                self.fallingPiece['rotation'] = (self.fallingPiece['rotation'] + 1) % len(PIECES[self.fallingPiece['shape']])
+
+        # move the current piece all the way down
+        elif (input[4] == 1):
+            self.movingDown = False
+            self.movingLeft = False
+            self.movingRight = False
+            for i in range(1, BOARDHEIGHT):
+                if not self.isValidPosition(adjY=i):
+                    break
+            self.fallingPiece['y'] += i - 1
+
+        # handle moving the piece because of user input
+        if (self.movingLeft or self.movingRight):
+            if self.movingLeft and self.isValidPosition(adjX=-1):
+                self.fallingPiece['x'] -= 1
+            elif self.movingRight and self.isValidPosition(adjX=1):
+                self.fallingPiece['x'] += 1
+            self.lastMoveSidewaysTime = time.time()
+
+        if self.movingDown:
+            self.fallingPiece['y'] += 1
+            self.lastMoveDownTime = time.time()
+
+        # let the piece fall if it is time to fall
+        # see if the piece has landed
+        cleared = 0
+        if not self.isValidPosition(adjY=1):
+            # falling piece has landed, set it on the self.board
+            self.addToBoard()
+
+            cleared = self.removeCompleteLines()
+            if cleared > 0:
+                if cleared == 1:
+                    self.score += 40 * self.level
+                elif cleared == 2:
+                    self.score += 100 * self.level
+                elif cleared == 3:
+                    self.score += 300 * self.level
+                elif cleared == 4:
+                    self.score += 1200 * self.level
+
+            self.score += self.fallingPiece['y']
+
+            self.lines += cleared
+            self.total_lines += cleared
+
+            reward = self.height - self.getHeight()
+            self.height = self.getHeight()
+
+            self.level, self.fallFreq = self.calculateLevelAndFallFreq()
+            self.fallingPiece = None
+
+        else:
+            # piece did not land, just move the piece down
+            self.fallingPiece['y'] += 1
+
+        # drawing everything on the screen
+        DISPLAYSURF.fill(BGCOLOR)
+        self.drawBoard()
+        #self.drawStatus()
+        #self.drawNextPiece()
+        if self.fallingPiece != None:
+           self.drawPiece(self.fallingPiece)
+
+        pygame.display.update()
+
+        if cleared > 0:
+            reward = 100 * cleared
+
+        image_data = pygame.surfarray.array3d(pygame.display.get_surface())
+        return image_data, reward, terminal
+
+    def getImage(self):
+        image_data = pygame.surfarray.array3d(pygame.display.get_surface())
+        return image_data
+
+    def getActionSet(self):
+        return range(6)
+
+    def getHeight(self):
```

Comment:
perhaps rename this to `getBoardHeight` to remove possible confusion. Initially thought this was returning the rendered screen height.